db/types/operations/cast.py

```diff
@@ -1027,6 +1036,128 @@ def _build_numeric_array_function():
     """
 
 
+
+
+
+def _build_integer_array_function():
+    """
+    The main reason for this function to be separate is for testing. This
+    does have some performance impact; we should consider inlining later.
+    """
+    qualified_function_name = get_qualified_name(INTEGER_ARR_FUNC_NAME)
+
+    no_separator_big = r"[0-9]{4,}(?:([,.])[0-9]+)?"
+    no_separator_small = r"[0-9]{1,3}(?:([,.])[0-9]{1,2}|[0-9]{4,})?"
+   
+    inner_number_tree = "|".join(
+        [
+            no_separator_big,
+            no_separator_small,
+            ])
+    integer_finding_regex = f"^(?:[+-]?({inner_number_tree}))$"
+
+    actual_number_indices = [1]
+    group_divider_indices = [4, 6, 8, 10, 12, 14, 16]
+    actual_numbers_str = ','.join([f'raw_arr[{idx}]' for idx in actual_number_indices])
+    group_dividers_str = ','.join([f'raw_arr[{idx}]' for idx in group_divider_indices])
+    text_db_type_id = PostgresType.TEXT.id
+    return rf"""
+    CREATE OR REPLACE FUNCTION {qualified_function_name}({text_db_type_id}) RETURNS {text_db_type_id}[]
+    AS $$
+      DECLARE
+        raw_arr {text_db_type_id}[];
+        actual_number_arr {text_db_type_id}[];
+        group_divider_arr {text_db_type_id}[];
+       
+        actual_number {text_db_type_id};
+        group_divider {text_db_type_id};
+       
+      BEGIN
+        SELECT regexp_matches($1, '{integer_finding_regex}') INTO raw_arr;
+        IF raw_arr IS NULL THEN
+          RETURN NULL;
+        END IF;
+        SELECT array_remove(ARRAY[{actual_numbers_str}], null) INTO actual_number_arr;
+        SELECT array_remove(ARRAY[{group_dividers_str}], null) INTO group_divider_arr;
+        SELECT actual_number_arr[1] INTO actual_number;
+        SELECT group_divider_arr[1] INTO group_divider;
+        RETURN ARRAY[actual_number, group_divider];
+      END;
+    $$ LANGUAGE plpgsql;
+    """
+
+
+def _build_decimal_array_function():
```

Comment:
The only place in the   _build_numeric_array_function(): where it being decimal or numeric will make a difference is in this line 
 qualified_function_name = get_qualified_name(NUMERIC_ARR_FUNC_NAME)
other than this it is independent ,so is this the only change required ? 

Also real and float types are a subset of numeric as given in the categories file ,so it should be taken care of in the numeric function itself 

if the same function is being used for both decimal and numeric types should the name of the function be changed ?

Reply:
I don't see any reason to consider `REAL` or `FLOAT as different from `NUMERIC` for this logic. The only possible issue is that these target types are of finite precision rather than arbitrary.

You should remove the distinction between them and just use the same casting function for all these types.

Reply:
So casting everything to numeric types instead of decimal types in the cast function should take care of it ?

def create_decimal_number_casts(engine):
    decimal_array_create = _build_numeric_array_function(DECIMAL_ARR_FUNC_NAME)
    with engine.begin() as conn:
        conn.execute(text(decimal_array_create))
    type_body_map = _get_decimal_number_type_body_map()
    create_cast_functions(PostgresType.NUMERIC, type_body_map, engine)